WcaOnRails/app/models/user.rb

```diff
@@ -1057,33 +1057,50 @@ def self.search(query, params: {})
     users.order(:name)
   end
 
+  def url
+    if wca_id
+      Rails.application.routes.url_helpers.person_url(wca_id, host: EnvVars.ROOT_URL)
+    else
+      ""
+    end
+  end
+
+  DEFAULT_SERIALIZE_OPTIONS = {
+    only: ["id", "wca_id", "name", "gender",
+           "country_iso2", "delegate_status", "created_at", "updated_at"],
+    methods: ["url"],
+    include: ["avatar", "teams"],
+  }.freeze
+
   def serializable_hash(options = nil)
-    json = {
-      class: self.class.to_s.downcase,
-      url: self.wca_id ? Rails.application.routes.url_helpers.person_url(self.wca_id, host: EnvVars.ROOT_URL) : "",
-
-      id: self.id,
-      wca_id: self.wca_id,
-      name: self.name,
-      gender: self.gender,
-      country_iso2: self.country_iso2,
-      delegate_status: delegate_status,
-      created_at: self.created_at,
-      updated_at: self.updated_at,
-      teams: current_team_members.includes(:team).reject do |team_member|
-        team_member.team.hidden?
-      end.map do |team_member|
-        {
-          friendly_id: team_member.team.friendly_id,
-          leader: team_member.team_leader?,
-        }
-      end,
-      avatar: {
+    # NOTE: doing deep_dup is necessary here to avoid changing the inner values
+    # of the freezed variables (which would leak PII)!
+    default_options = DEFAULT_SERIALIZE_OPTIONS.deep_dup
+    # Delegates's emails and regions are public information.
+    if any_kind_of_delegate?
+      default_options[:methods].concat(["email", "region", "senior_delegate_id"])
+    end
+
+    options = default_options.merge(options || {})
+    # Preempt the values for avatar and teams, they have a special treatment.
+    include_avatar = options[:include]&.delete("avatar")
+    include_teams = options[:include]&.delete("teams")
```

Comment:
`:include` is a valid key of the `default_options` hash, so I don't see how it would ever be `nil` here.
Can `.merge()` above delete keys?!

Reply:
`.merge()` cannot delete keys, but you can very well nullify it in the option with something like this:
```ruby
irb(main):003:0> User.last.as_json(include: nil, only: ["name", "id", "url"])
=> {"id"=>1234, "name"=>"redacted", "url"=>"", "class"=>"user"}
```
To only get very basic data about the user.

Reply:
Ah. Of course, adding `nil` explicitly is always a possibility which I didn't consider. Point well taken :smile: